x-pack/plugins/licensing/server/license.ts

```diff
@@ -0,0 +1,178 @@
+    return this.hasLicense;
+  }
+
+  public get isBasic() {
+    return this.isActive && this.type === 'basic';
+  }
+
+  public get isNotBasic() {
+    return this.isActive && this.type !== 'basic';
+  }
+
+  public get reasonUnavailable() {
+    if (!this.isAvailable) {
+      return `[${this.clusterSource}] Elasticsearch cluster did not respond with license information.`;
+    }
+
+    if (this.error instanceof Error && (this.error as any).status === 400) {
+      return `X-Pack plugin is not installed on the [${this.clusterSource}] Elasticsearch cluster.`;
+    }
+
+    return this.error;
+  }
+
+  public get signature() {
+    if (this._signature !== undefined) {
+      return this._signature;
+    }
+
+    this._signature = createHash('md5')
+      .update(JSON.stringify(this.toObject()))
+      .digest('hex');
+
+    return this._signature;
+  }
+
+  isOneOf(candidateLicenses: string | string[]) {
+    if (!Array.isArray(candidateLicenses)) {
+      candidateLicenses = [candidateLicenses];
+    }
+
+    return candidateLicenses.includes(this.type);
+  }
+
+  meetsMinimumOf(minimum: LICENSE_TYPE) {
+    return LICENSE_TYPE[this.type as LicenseType] >= minimum;
+  }
+
+  check(pluginName: string, minimumLicenseRequired: LICENSE_TYPE | string) {
+    const minimum = toLicenseType(minimumLicenseRequired);
+
+    if (!this.isAvailable) {
+      return {
+        check: LICENSE_STATUS.Unavailable,
+        message: i18n.translate('xpack.licensing.check.errorUnavailableMessage', {
+          defaultMessage:
+            'You cannot use {pluginName} because license information is not available at this time.',
+          values: { pluginName },
+        }),
+      };
+    }
+
+    const { type: licenseType } = this.license;
+
+    if (!this.meetsMinimumOf(minimum)) {
+      return {
+        check: LICENSE_STATUS.Invalid,
+        message: i18n.translate('xpack.licensing.check.errorUnsupportedMessage', {
+          defaultMessage:
+            'Your {licenseType} license does not support {pluginName}. Please upgrade your license.',
+          values: { licenseType, pluginName },
+        }),
+      };
+    }
+
+    if (!this.isActive) {
+      return {
+        check: LICENSE_STATUS.Expired,
+        message: i18n.translate('xpack.licensing.check.errorExpiredMessage', {
+          defaultMessage:
+            'You cannot use {pluginName} because your {licenseType} license has expired.',
+          values: { licenseType, pluginName },
+        }),
+      };
+    }
+
+    return { check: LICENSE_STATUS.Valid };
+  }
+
+  toObject() {
+    if (this.objectified) {
+      return this.objectified;
+    }
+
+    this.objectified = {
+      license: {
+        type: this.type,
+        isActive: this.isActive,
+        expiryDateInMillis: this.expiryDateInMillis,
+      },
+      features: [...this.featuresMap].map(([, feature]) => feature.toObject()),
+    };
+
+    return this.objectified;
+  }
+
+  getFeature(name: string) {
```

Comment:
with immutable `License` class, `LicenseFeature` doesn't track license updates

Reply:
The `LicenseFeature` class isn't meant to track changes. That is the responsibility of the feature plugin.